### PR TITLE
[http_endpoint input] Add support for including headers and preserving original events

### PR DESCRIPTION
**Pull request by P1llus**
## What does this PR do?

This adds the possibility to give a list of headers that, if they exist in the incoming request, will be added as a separate object. Also adds the possibility to preserve the raw event as part of event.original.

Example output here:

Config example:
```
- type: http_endpoint
  enabled: true
  listen_address: 10.10.10.10
  listen_port: 8080
  preserve_original_event: true
  include_headers: ["TestHeader"]
```

```
  "json": {
    "test": "somevalue"
  },
  "event": {
    "original": "{\"test\":\"somevalue\"}"
  },
  "headers": {
    "TestHeader": "test"
  }
```

## Why is it important?

Certain modules is hard to parse data without the extra details from the headers, github webhooks is one example of that, in which the data type is stored in the header

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: x-pack/filebeat/input/http_endpoint/handler.go [hunk 1]
```diff
@@ -113,3 +124,14 @@ func httpReadJSON(body io.Reader) (objs []common.MapStr, status int, err error)
 	}
 	return objs, 0, nil
 }
+
+func getIncludedHeaders(r *http.Request, headerconf []string) (includedHeaders common.MapStr) {
+	includedHeaders = common.MapStr{}
+	for _, header := range headerconf {
+		h := r.Header.Get(header)
```
Comment (marc-gr) [inline x-pack/filebeat/input/http_endpoint/handler.go:131]: Maybe accessing the map directly would be better for the purpose here, since `Get` only returns the first value for the header, but http headers can contain multiple (https://golang.org/pkg/net/http/#Header.Get).

```suggestion
		h := r.Header[textproto. CanonicalMIMEHeaderKey(header)] // CanonicalMIMEHeaderKey will set the right casing to get the header, Get does this automatically
```

Comment (P1llus) [reply x-pack/filebeat/input/http_endpoint/handler.go]: Applied the changes, will have to take into account that the value of the header will now be arrays, but that should still work.

File: x-pack/filebeat/input/http_endpoint/handler.go [hunk 2]
```diff
@@ -53,13 +59,19 @@ func (h *httpHandler) sendResponse(w http.ResponseWriter, status int, message st
...
-func (h *httpHandler) publishEvent(obj common.MapStr) {
+func (h *httpHandler) publishEvent(obj common.MapStr, headers common.MapStr) {
 	event := beat.Event{
 		Timestamp: time.Now().UTC(),
 		Fields: common.MapStr{
 			h.messageField: obj,
 		},
 	}
+	if h.preserveOriginalEvent {
+		event.PutValue("event.original", obj.String())
```
Comment (andrewkroh) [inline x-pack/filebeat/input/http_endpoint/handler.go:70]: This is reconstructing the JSON by marshaling it again. That's not really the true original bytes received in the request. To get the raw bytes you would want to change `httpReadJSON` to `Decode` into a [`json.RawMessage`](https://golang.org/pkg/encoding/json/#RawMessage) so that you can defer the unmarshalling until after you've saved a copy the raw JSON bytes.

**Review by andrewkroh — CHANGES_REQUESTED** [review]

File: x-pack/filebeat/tests/system/test_http_endpoint.py [hunk 3]
```diff
@@ -308,6 +308,50 @@ def test_http_endpoint_invalid_hmac(self):
...
+  preserve_original_event: true
+"""
+        self.get_config(options)
+        filebeat = self.start_beat()
+        self.wait_until(lambda: self.log_contains("Starting HTTP server on {}:{}".format(self.host, self.port)))
+
+        message = "somerandommessage"
+        payload = {self.prefix: message}
+        headers = {"Content-Type": "application/json"}
+        r = requests.post(self.url, headers=headers, data=json.dumps(payload))
```
[the remark this answers is not in the document]
Comment (andrewkroh) [reply x-pack/filebeat/tests/system/test_http_endpoint.py:325]: You might even want to make the `json.dumps` pretty with an `indent` so that you can verify that you are getting the exact bytes back and not something that was reconstructed (assuming that's the requirement you are fulfilling).

File: x-pack/filebeat/tests/system/test_http_endpoint.py [hunk 4]
```diff
@@ -308,6 +308,50 @@ def test_http_endpoint_invalid_hmac(self):
...
+        options = """
+  include_headers: ["TestHeader"]
+"""
+        self.get_config(options)
+        filebeat = self.start_beat()
+        self.wait_until(lambda: self.log_contains("Starting HTTP server on {}:{}".format(self.host, self.port)))
+
+        message = "somerandommessage"
+        payload = {self.prefix: message}
+        headers = {"Content-Type": "application/json", "TestHeader": "TestHeaderValue"}
```
[the remark this answers is not in the document]
Comment (P1llus) [reply x-pack/filebeat/tests/system/test_http_endpoint.py:346]: Multiple headers is not possible unfortunately, its not really support by python requests itself, so no matter what way you want to craft a multi-value header, it will deny it, if you know a workaround I would be happy to implement it.

File: x-pack/filebeat/input/http_endpoint/handler.go (hunk 2 above)
Comment (P1llus) [reply x-pack/filebeat/input/http_endpoint/handler.go:70]: This one would be a bit tough, because this is after iterating over multiple objects in a single incoming request, we don't want to include the whole event.original in each of the objects, I might contact you to discuss this one a bit more

File: x-pack/filebeat/tests/system/test_http_endpoint.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (P1llus) [reply x-pack/filebeat/tests/system/test_http_endpoint.py:352]: Ref for multi-header: 
https://docs.python.org/3/library/urllib.request.html#urllib.request.Request.add_header
" Note that there cannot be more than one header with the same name, and later calls will overwrite previous calls in case the key collides"

File: x-pack/filebeat/input/http_endpoint/handler.go [hunk 5]
```diff
@@ -82,34 +93,94 @@ func sendErrorResponse(w http.ResponseWriter, status int, err error) {
...
+			}
+			return nil, nil, errors.Wrapf(err, "malformed JSON object at stream position %d", dec.InputOffset())
+		}
+
+		var obj interface{}
+		if err := newJSONDecoder(bytes.NewReader(raw)).Decode(&obj); err != nil {
+			return nil, nil, errors.Wrapf(err, "malformed JSON object at stream position %d", dec.InputOffset())
+		}
+
+		rawMessages = append(rawMessages, raw)
```
Comment (andrewkroh) [inline x-pack/filebeat/input/http_endpoint/handler.go:162]: To be sure the size of the returned slices are always the same I think this should be moved down to where `objs` is appended to.

File: x-pack/filebeat/tests/system/test_http_endpoint.py [hunk 6]
```diff
@@ -308,6 +308,144 @@ def test_http_endpoint_invalid_hmac(self):
...
+        message = "somerandommessage"
+        payload = {self.prefix: message}
+        bp = json.dumps(payload, indent=2).encode('utf-8')
+        headers = {"Content-Type": "application/json"}
+        r = requests.post(self.url, headers=headers, data=json.dumps(payload))
+
+        filebeat.check_kill_and_wait()
+        output = self.read_output()
+
+        # To be able to do an exact comparison, it needs to be converted back into a dict, before reading it as bytes
```
Comment (andrewkroh) [inline x-pack/filebeat/tests/system/test_http_endpoint.py:331]: I think the value in this test is that it verifies that `event.original` matches the raw data that was sent (ignoring leading and trailing whitespace). So if the test unmarshals `event.original` and marshals it again then compares that to the original payload bytes then we haven't really varied that we preserved the source content.

What are the whitespace issues that required this? Can you trim the leading/trailing to remove possible newline issues and see if that works?

**Review by andrewkroh — APPROVED** [review]
